schematic/loader.py

```diff
@@ -0,0 +1,103 @@
+from errno import ENOENT
+from os import pathsep
+from re import split
+from pkg_resources import resource_exists, \
+  resource_filename, \
+  resource_stream, \
+  resource_string, \
+  resource_listdir
+
+class InvalidResourceError(Exception):
+    """
+    Args:
+        uri {String}: The URI which was requested within the given loader's 
+        that did not exist or was malformed.
+    """  
+    def __init__(self, namespace, requested_uri):
+        self.namespace = namespace
+        self.requested_uri = requested_uri
+        self.message = 'Resource does not exist or is declared incorrectly'
+        self.errno = ENOENT
+        super(InvalidResourceError, self).__init__(self.message)
+
+
+    def __str__(self):
+        return '{}({}), "{}" of {}'.format(self.message, self.errno, self.requested_uri, self.namespace)
+
+
+    def __repr__(self):
+        return self.__str__()
+
+        
+class Loader(object):
```

Comment:
Where was this `Loader` pattern taken from? This seems like a lot of code for just loading files from within the package.

Reply:
It's not a standard pattern, it was inspired by an example I found which showed the use of `setuptools` / `pkg_resources`. So `Loader()` is a library, there are many ways to load package resources -- `resource_string`, `resource_stream`, etc. so there are wrappers for each of those methods. If someone requires say, loading all the contents from a package resource in the future then they can just use `resource_string` wrapper (as `Loader.read()`) rather than relying on `Loader.filename()` (which uses `resource_stream()`).